### PR TITLE
fix(team-task-board): harden IPC validation and attachment rollback

### PR DESCRIPTION
**Pull request by 777genius**
## Summary

- validate every blockedBy and orderedTaskIds entry atomically at the IPC boundary
- enforce one 20 MiB decoded attachment policy and reject oversized encoded payloads before decoding or persistence
- roll back already-saved attachments when a later save or comment write fails
- keep rollback best-effort while preserving the original persistence error

## Root cause

The extracted task-board handlers validated collection shapes but did not validate every task ID. Attachment size enforcement existed only in storage, and multi-step comment persistence had no compensating cleanup after partial failure.

## Impact

Malformed order/dependency payloads no longer reach application ports. Failed multi-attachment comments no longer leave successfully saved earlier files behind under normal adapter behavior.

## Verification

- `pnpm exec vitest run src/features/team-task-board/main/adapters/input/ipc/registerTeamTaskBoardIpc.test.ts src/features/team-task-board/main/composition/createTeamTaskBoardFeature.test.ts src/features/team-task-board/core/application/use-cases/UpdateTaskFieldsUseCase.test.ts test/features/team-task-board/core/domain/taskAttachmentPayloadPolicy.test.ts`
- `pnpm typecheck`
- `pnpm lint:fast:files -- <12 changed files>`
- `pnpm guard:source-file-size`

## Integration gate

- Draft only
- Base snapshot: `082bf7e8fd426578905efaab97645bd6ef98b31c`
- Do not merge until the canonical serial gate authorizes this PR
- No runtime, desktop E2E, Changes Page E2E, or real-project execution was used

## Summary by CodeRabbit

* **New Features**
  * Added shared task-attachment payload limits, canonical ID/base64 validation, and decoded-size estimation utilities.
  * Introduced an attachment-aware add-task-comment workflow with saved-attachment rollback support.
* **Bug Fixes**
  * Strengthened IPC input validation for `blockedBy` and kanban ordering IDs (rejects invalid entries before execution).
  * Improved attachment persistence/cleanup durability and ensured safe behavior when atomic directory sync partially fails.
* **Tests**
  * Expanded policy, IPC, store, and use-case coverage for boundary cases, duplicates, invalid inputs, and rollback ordering/cleanup.

### CODE REVIEW

**Comment by 777genius** [comment]
Final exact-head hosted review complete for e7dbbc2d27c30166365a33cb6316369022e5bacc against base 082bf7e8fd426578905efaab97645bd6ef98b31c.

Result: NO P0-P3 findings.

Hosted proof:
- double-EBUSY real-filesystem delete and rollback both remove the published target and owned temp hardlink
- 50 repeated delete/save/stale-rollback races: 0 replacement deletions, 0 path reuse, 0 wrong bytes
- 107/107 focused tests
- typecheck, architecture, exact production lint, fast lint, and 800-line guard pass
- all GitHub checks are green, including both test shards, CodeQL, validate, lints, and Windows smoke

The PR remains draft and unmerged. No canonical or dev branch was changed.

**Comment by 777genius** [comment]
@codex review\n\nPlease review exact head 7fbf4d99bce5657288047bc13b3e0bc9d1b21692. Focus on cross-process attachment ID uniqueness, atomic no-clobber publication, rollback identity safety, and legacy attachment compatibility.

**Comment by 777genius** [comment]
@codex review

**Comment by 777genius** [comment]
@codex review\n\nExact head: bc98442fe62273b239ad3d4d5afb02a869e40494\n\nPlease review the exact head. Focus on cross-process task attachment locking, generation guard finalization, rollback identity safety, cleanup races, and metadata/comment commit ordering.

**Comment by 777genius** [comment]
@codex review\n\nPlease review exact head 4a3a19a5e after canonical d4cb9a6d synchronization. Focus on attachment generation finalization, cross-process task-directory locking, rollback boundaries, and race coverage.

**Comment by 777genius** [comment]
@codex review

Please review exact head 07b9dbde3. Focus on the task-scoped unit of work across attachment publication, comment/metadata persistence, finalization, rollback and deletion; compromised/release lock outcomes; compensation; and backup/restore filtering of generation guards.

**Comment by 777genius** [comment]
@codex review

Fresh exact head: 210c3022b0c23e38407ddc7e8ae163326944bf13

Please re-review the complete PR diff against canonical a12b030f884cb3b9c94ba6cf609f1322257126a7, with focus on attachment generation identity, crash-safe delete ordering, transaction commit and rollback, and backup artifact pruning.

**Comment by 777genius** [comment]
@codex review

File: src/main/services/team/TeamTaskAttachmentStore.ts [hunk 1]
```diff
@@ -94,41 +118,137 @@ export class TeamTaskAttachmentStore {
...
+        receipt.identity.ino === 0 ||
+        stats.dev !== receipt.identity.dev ||
+        stats.ino !== receipt.identity.ino
+      ) {
+        return;
+      }
+
+      await this.atomicCreator.cleanupPublishedTempLinks(filePath);
+      try {
+        await fs.promises.unlink(filePath);
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/TeamTaskAttachmentStore.ts]: Fixed with cross-process task transactions and generation-guarded receipts. Exact-head tests cover delete and recreate races.

File: src/main/services/team/TeamTaskAttachmentStore.ts [hunk 2]
```diff
@@ -94,41 +118,137 @@ export class TeamTaskAttachmentStore {
...
-    const trimmed = base64Data.trim();
+    const receipt = await this.saveAttachmentWithReceipt(
+      teamName,
+      taskId,
+      attachmentId,
+      filename,
+      mimeType,
+      base64Data
+    );
+    return receipt.metadata;
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/TeamTaskAttachmentStore.ts]: Fixed. The save flow retains the receipt and rolls back the exact generation when metadata persistence fails.

File: src/main/services/team/TeamTaskAttachmentStore.ts [hunk 3]
```diff
@@ -94,41 +156,185 @@ export class TeamTaskAttachmentStore {
...
+      }
 
-    const filePath = this.getStoredFilePath(teamName, taskId, attachmentId, filename);
-    await fs.promises.writeFile(filePath, buffer);
+      const filePath = this.getStoredFilePath(teamName, taskId, attachmentId);
+      guard.assertHealthy();
+      const identity = await this.atomicCreator.createFileAtomically(filePath, buffer);
+      let generationGuardPath: string | undefined;
+      try {
+        generationGuardPath = await this.atomicCreator.createGenerationGuard?.(filePath);
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/TeamTaskAttachmentStore.ts]: Fixed. Successful persistence finalizes the receipt and removes its generation guard; interrupted cleanup is recoverable.

File: src/main/services/team/TeamTaskAttachmentStore.ts [hunk 4]
```diff
@@ -163,27 +369,36 @@ export class TeamTaskAttachmentStore {
...
+  private async cleanupEmptyTaskDirectory(teamName: string, taskId: string): Promise<void> {
     const dir = this.getTaskDir(teamName, taskId);
     try {
-      const entries = await fs.promises.readdir(dir);
-      if (entries.length === 0) {
-        await fs.promises.rm(dir, { recursive: true });
-      }
-    } catch {
-      // ignore cleanup errors
+      await fs.promises.rmdir(dir);
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/TeamTaskAttachmentStore.ts:772]: Fixed. Save and delete cleanup now coordinate at task scope, including the last-file directory race.

File: src/features/team-task-board/core/application/use-cases/AddTaskCommentUseCase.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+          taskId,
+          attachment.id,
+          attachment.filename,
+          attachment.mimeType,
+          attachment.base64Data
+        );
+        savedAttachments.push(saved);
+      }
+
+      comment = await this.dependencies.comments.addTaskComment(
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/features/team-task-board/core/application/use-cases/AddTaskCommentUseCase.ts:74]: Fixed. Attachment publication, comment persistence, finalization, and rollback now run inside the same task transaction.

File: src/main/services/team/TaskAttachmentMutationCoordinator.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      });
+      const guard: TaskAttachmentMutationGuard = {
+        assertHealthy() {
+          if (compromisedError) throw compromisedError;
+        },
+      };
+
+      try {
+        guard.assertHealthy();
+        return await operation(guard);
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/TaskAttachmentMutationCoordinator.ts]: Fixed. The coordinator rechecks compromise after the operation and compensates only uncommitted publication.

File: src/features/team-task-board/core/application/use-cases/AddTaskCommentUseCase.ts [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+          ? savedAttachments.map((attachment) => attachment.metadata)
+          : undefined,
+        input.taskRefs
+      );
+    } catch (error) {
+      await this.rollbackSavedAttachments(savedAttachments);
+      throw error;
+    }
+
+    await this.finalizeSavedAttachments(savedAttachments);
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/features/team-task-board/core/application/use-cases/AddTaskCommentUseCase.ts:89]: Fixed. Durable lifecycle recovery reconciles interrupted guard finalization, and backups exclude internal guard payloads.

File: src/main/services/team/TaskAttachmentMutationCoordinator.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+      } finally {
+        await release();
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/TaskAttachmentMutationCoordinator.ts]: Fixed. A committed mutation is not reclassified as failed when lock release cleanup fails.

File: src/main/services/team/TeamTaskAttachmentStore.ts [hunk 9]
```diff
@@ -121,41 +178,222 @@ export class TeamTaskAttachmentStore {
...
+          if (generationGuardPath) {
+            await fs.promises.unlink(generationGuardPath).catch(() => undefined);
+          }
+          return;
+        }
 
-    logger.debug(`[${teamName}] Saved task attachment ${attachmentId} for task #${taskId}`);
-    return meta;
+        guard.assertHealthy();
+        await this.atomicCreator.cleanupPublishedTempLinks(filePath);
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/TeamTaskAttachmentStore.ts]: Fixed. Identity-fenced target rollback is attempted independently from temporary-link cleanup.

File: src/main/services/team/TeamTaskAttachmentStore.ts [hunk 10]
```diff
@@ -121,41 +259,317 @@ export class TeamTaskAttachmentStore {
...
-    const meta: TaskAttachmentMeta = {
+    const filePath = this.getStoredFilePath(teamName, taskId, attachmentId);
+    guard.assertHealthy();
+    const identity = await this.atomicCreator.createFileAtomically(filePath, buffer);
+    let generationGuardPath: string | undefined;
+    try {
+      generationGuardPath = await this.atomicCreator.createGenerationGuard?.(filePath);
+    } catch (error) {
+      await this.atomicCreator.cleanupPublishedTempLinks(filePath).catch(() => undefined);
+      await fs.promises.unlink(filePath).catch(() => undefined);
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/TeamTaskAttachmentStore.ts]: Fixed. Failure cleanup is fenced by the captured generation identity and cannot unlink a replacement.

File: src/main/services/team/TaskAttachmentMutationCoordinator.ts [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+              entry.active = false;
+            },
+          };
+        },
+      };
+
+      try {
+        guard.assertHealthy();
+        const result = await operation(guard);
+        guard.assertHealthy();
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/TaskAttachmentMutationCoordinator.ts:83]: Fixed. The coordinator tracks committed terminal state, so a late compromise cannot report a false failure.

File: src/main/services/team/TeamBackupService.ts [hunk 12]
```diff
@@ -955,6 +943,7 @@ export class TeamBackupService {
         for (const entry of entries) {
           const relPath = prefix ? `${prefix}/${entry.name}` : entry.name;
           if (entry.isFile()) {
+            if (!shouldCollectTaskAttachmentBackupPath(relPath)) continue;
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/TeamBackupService.ts]: Fixed. Restore filtering is separate from maintenance enumeration, so stale internal backup paths remain prunable.

File: src/main/services/team/TaskAttachmentGenerationLifecycle.ts [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+export type PinTaskAttachmentGenerationResult =
+  | { readonly kind: 'pinned'; readonly receipt: PinnedTaskAttachmentGeneration }
+  | { readonly kind: 'missing' }
+  | { readonly kind: 'changed' };
+
+function isSameIdentity(
+  left: Pick<fs.Stats, 'dev' | 'ino'>,
+  right: TaskAttachmentFileIdentity
+): boolean {
+  return left.dev === right.dev && left.ino !== 0 && left.ino === right.ino;
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/services/team/TaskAttachmentGenerationLifecycle.ts]: Fixed. An unsupported zero inode is rejected before public publication or pin creation, with clean-retry coverage.

File: src/main/ipc/teams.ts [hunk 14]
```diff
@@ -729,13 +750,30 @@ async function handleDeleteTaskAttachment(
...
+                `[teams:deleteTaskAttachment] Failed to restore attachment ${safeAttId}: ${
+                  rollbackError instanceof Error ? rollbackError.message : String(rollbackError)
+                }`
+              );
+            }
+          }
+          throw error;
+        }
+        transaction.markCommitted();
+        if (receipt) await transaction.finalizeAttachmentDeletion(receipt);
```
[the remark this answers is not in the document]
Comment (777genius) [reply src/main/ipc/teams.ts:775]: Fixed. Post-commit attachment deletions are durably journaled and reconciled after restart or transient failure.

**Comment by 777genius** [comment]
@codex review

**Comment by 777genius** [comment]
@codex review

Exact head: 5791588ac1d636f0d0b84fe0c4b07f565dbf8694
All CI checks are green. Please review this exact revision and include the reviewed commit in the verdict.

**Comment by 777genius** [comment]
@codex review

Exact head: e809547b21d0f4d4a0c6cf8c3803d434b13d7cc8
Canonical base: c84fb2232a21b8d3911989b2629eaf81aa8f44c0
Please review this exact revision and report any P0-P2 findings.

**Comment by 777genius** [comment]
@codex review

Exact head: 0619618200a8317872792c4c82413bb8b91b44a4
Please review the remediation delta from e809547b21d0f4d4a0c6cf8c3803d434b13d7cc8. Focus on TEAM_CREATE_TASK related ID validation, whitespace normalization, persisted relationship identity, undefined and empty-array semantics, and the regression test.
